modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,424 @@
+     */
+    private final ClusterService clusterService;
+
+    /**
+     * Folder for units.
+     */
+    private Path unitsFolder;
+
+    /**
+     * Constructor.
+     *
+     * @param clusterService Cluster service.
+     * @param metaStorage Meta storage.
+     * @param workDir Node working directory.
+     * @param configuration Deployment configuration.
+     * @param cmgManager Cluster management group manager.
+     */
+    public DeploymentManagerImpl(ClusterService clusterService,
+            MetaStorageManager metaStorage,
+            Path workDir,
+            DeploymentConfiguration configuration,
+            ClusterManagementGroupManager cmgManager) {
+        this.clusterService = clusterService;
+        this.metaStorage = metaStorage;
+        this.configuration = configuration;
+        this.cmgManager = cmgManager;
+        unitsFolder = workDir;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> deployAsync(String id, Version version, DeploymentUnit deploymentUnit) {
+        if (id == null || id.isBlank() || version == null) {
+            return CompletableFuture.failedFuture(new DeploymentUnitIdentifierException());
+        }
+
+        ByteArray key = new ByteArray(UNITS_PREFIX + id + ":" + version.render());
+
+        UnitMeta meta = new UnitMeta(id, version, deploymentUnit.name(), Collections.emptyList());
+
+        Operation put = put(key, UnitMetaSerializer.serialize(meta));
+
+        DeployUnitRequestBuilder builder = DeployUnitRequestImpl.builder();
+
+        try {
+            builder.unitContent(deploymentUnit.content().readAllBytes());
+        } catch (IOException e) {
+            LOG.error("Error to read deployment unit content", e);
+            return CompletableFuture.failedFuture(new DeploymentUnitReadException(e));
+        }
+        DeployUnitRequest request = builder
+                .unitName(deploymentUnit.name())
+                .id(id)
+                .version(version.render())
+                .build();
+
+        return metaStorage.invoke(notExists(key), put, Operations.noop())
+                .thenCompose(success -> {
+                    if (success) {
+                        return doDeploy(request);
+                    }
+                    LOG.error("Failed to deploy meta of unit " + id + ":" + version);
+                    return CompletableFuture.failedFuture(new DeployUnitWriteMetaException());
+                })
+                .thenApply(completed -> {
+                    if (completed) {
+                        startDeployAsyncToCmg(request);
+                    }
+                    return completed;
+                });
+    }
+
+    private void startDeployAsyncToCmg(DeployUnitRequest request) {
+        cmgManager.cmgNodes()
+                .thenApply(nodes -> nodes.stream().map(node -> clusterService.topologyService().getByConsistentId(node)).collect(
+                        Collectors.toList()))
+                .thenAccept(clusterNodes -> {
```

Comment:
Ok, so you create a list of node, and the only usage for it is "iteration".
Streams were designed with other intention in mind. For example, you can use "forEach", in that case no new list would be created, making you code faster and more "idiomatic", I guess.

Maybe there are hidden reasons for having an explicit list, that I can't understand right now. In that case, please express your reasoning in the comment.